### PR TITLE
perf(hooks): optimize formatter hooks(x52 faster) — local binary, merged invocations, direct require()

### PR DESCRIPTION
**Pull request by pythonstrup**
## Description

Optimize the post-edit-format and quality-gate hooks for significantly faster execution by eliminating `npx` overhead, merging redundant Biome invocations, and reducing the process spawn chain.

### Benchmark Results

| Metric | Before (npx) | After (local binary) | Improvement |
|--------|-------------|---------------------|-------------|
| Biome format per Edit | ~3,300ms | ~63ms | **~52x faster** |
| Biome duplicate check (quality-gate) | ~3,300ms | 0ms (skipped) | **eliminated** |
| Process spawn chain | 3 (run-with-flags → node → npx → biome) | 1 (run-with-flags → biome) | **-2 spawns** |

### Changes

1. **Shared `resolve-formatter.js` utility** (`scripts/lib/resolve-formatter.js` — new)
   - Extracts `findProjectRoot()`, `detectFormatter()`, `resolveFormatterBin()` into a shared module
   - Caches results per-process to avoid redundant filesystem lookups
   - Resolves local `node_modules/.bin/biome` or `node_modules/.bin/prettier` binary first, falls back to `npx`

2. **Eliminate npx overhead** (`scripts/hooks/post-edit-format.js`)
   - Uses local binary directly instead of `npx` (~200-500ms savings per invocation)
   - Changes `biome format --write` → `biome check --write` (format + lint in one pass)
   - Exports `run()` function for direct invocation

3. **Skip redundant quality-gate checks** (`scripts/hooks/quality-gate.js`)
   - Skips Biome check for JS/TS files already handled by post-edit-format
   - Still runs Biome for `.json`/`.md` files
   - Fixes bug: uses `findProjectRoot()` instead of `process.cwd()` for config detection
   - Adds `timeout: 15000` to prevent hanging

4. **Direct `require()` invocation** (`scripts/hooks/run-with-flags.js`)
   - When hook exports `run()`, calls it directly via `require()` instead of `spawnSync('node', ...)`
   - Falls back to legacy spawnSync for hooks without `run()` export
   - Adds path traversal guard before `require()`
   - Adds `timeout: 30000` to legacy spawnSync path
   - Logs `require()` errors to stderr instead of silently swallowing

## Type of Change
- [x] `feat:` New feature
- [x] `fix:` Bug fix

## Testing

**Unit tests:**
- 15 new tests for `resolve-formatter.js` — all pass
- 190 existing hook tests — all pass
- 23 integration tests — all pass

**Live E2E tests with real projects:**

| Project | Formatter | Quotes | Semicolons | Indent | Result |
|---------|-----------|--------|------------|--------|--------|
| Biome (tab + double quote + semi) | Biome local | `"` ✓ | yes ✓ | tab ✓ | **PASS** |
| Prettier (4-space + single quote + no semi) | Prettier local | `'` ✓ | no ✓ | 4-space ✓ | **PASS** |
| No formatter | none | unchanged ✓ | unchanged ✓ | unchanged ✓ | **PASS** |

**26 comprehensive test cases covering:**
- Happy paths: Biome/Prettier local binary, npx fallback, 3 prettier config formats
- Edge cases: non-JS/TS files, missing files, empty files, paths with spaces, monorepo nested package.json, syntax errors, malformed JSON input
- Security: path traversal, shell metacharacter injection

## Checklist
- [x] Tests pass locally (`node tests/run-all.js`)
- [x] Validation scripts pass
- [x] Follows conventional commits format
- [x] Backwards compatible — Prettier fallback and stdin-based entry points preserved

---
## Summary by cubic
Speed up formatter hooks by using local binaries, merging `biome` runs, and invoking hooks in-process. Format+lint after edits now runs ~52x faster (≈3300ms → ~63ms), skips the redundant `biome` JS/TS check in the quality gate, and adds guarded Windows `.cmd` support.

- **Refactors**
  - Added `scripts/lib/resolve-formatter.js` with caching and shared constants: finds project root, detects `biome`/`prettier` (incl. `package.json` "prettier"), prefers local `node_modules/.bin`, and falls back to the project package manager runner (respects `CLAUDE_PACKAGE_MANAGER`, Windows `.cmd`; returns null for unknown).
  - Updated `post-edit-format` and `quality-gate`: use local bins; switch to `biome check --write` (one pass); skip JS/TS in quality gate when Biome is configured while still handling `.json`/`.md`, `prettier`/Go/Python; both export `run()` for direct invocation with `findProjectRoot`-based root discovery.

- **Bug Fixes**
  - Windows: `.cmd` shims executed via `spawnSync` with `shell:true`, gated by an unsafe-character filter to prevent injection.
  - `run-with-flags`: only `require()` hooks that export `run()`; add path containment guard, separate `require()`/`run()` error logging with pass-through on errors, and timeouts (15s in quality-gate exec, 30s for legacy spawn).

<sup>Written for commit d239d873d88717f1746747248a4327488881ff2d. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Hooks now expose an in-process run(rawInput) API while keeping existing CLI stdin/stdout behavior.
  * Quality gate adds Go and Python support with unified check/fix flows.

* **Refactor**
  * Centralized formatter detection/resolution with caching and unified execution.
  * Safer hook loading with absolute-path checks and graceful fallback to legacy execution.

* **Tests**
  * Comprehensive test coverage for detection, resolution, caching, and fallbacks.

### CODE REVIEW

File: scripts/lib/resolve-formatter.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    if (fs.existsSync(path.join(dir, 'package.json'))) {
+      projectRootCache.set(startDir, dir);
+      return dir;
+    }
+    dir = path.dirname(dir);
+  }
+
+  projectRootCache.set(startDir, startDir);
+  return startDir;
+}
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/lib/resolve-formatter.js:74]: Acknowledged — valid observation, but out of scope for this PR.

This PR focuses on **performance optimization** (eliminating npx overhead, merging invocations, direct require). The `findProjectRoot()` anchoring on `package.json` is pre-existing behavior unchanged by this PR, and the vast majority of JS/TS projects have a `package.json` at the root.

Adding formatter config file detection to `findProjectRoot()` would broaden its responsibility and introduce a behavioral change beyond the optimization scope. Tracking as a separate enhancement.

File: scripts/lib/resolve-formatter.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  if (formatter === 'prettier') {
+    const localBin = path.join(projectRoot, 'node_modules', '.bin', isWin ? 'prettier.cmd' : 'prettier');
+    if (fs.existsSync(localBin)) {
+      const result = { bin: localBin, prefix: [] };
+      binCache.set(cacheKey, result);
+      return result;
+    }
+    const result = { bin: npxBin, prefix: ['prettier'] };
+    binCache.set(cacheKey, result);
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/lib/resolve-formatter.js]: Acknowledged — valid point, but out of scope for this PR.

The npx fallback is pre-existing behavior (the original code also used `npx`). This PR's goal is to **prefer local binaries** over npx, which already avoids the npx path for most setups. Integrating `CLAUDE_PACKAGE_MANAGER` into the fallback runner is a separate concern and would touch the package-manager resolution layer. Tracking as a follow-up issue.

File: scripts/lib/resolve-formatter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/lib/resolve-formatter.js:74]: Yes, please open a follow-up issue for this. Thanks\!

File: scripts/lib/resolve-formatter.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/lib/resolve-formatter.js]: Yes, please open a follow-up issue for this as well. Thanks\!

**Review by affaan-m — CHANGES_REQUESTED** [review]
Automated review: this PR has merge conflicts. Please rebase or resolve.

**Comment by affaan-m** [comment]
This PR has merge conflicts. Could you rebase on the latest main when you get a chance? Thanks!

**Review by affaan-m — CHANGES_REQUESTED** [review]
Automated review: this PR has merge conflicts. Please rebase or resolve.
